### ISSUE TITLE
Fedora packaging thread

### ISSUE THREAD

**Issue by asymptotically** [facebook/watchman#474]
Hi, it would be nice if the configure script gave the option to use and dynamically link with system versions of libart, wildmatch, tap and the json libraries instead of building them here and using them statically.

Thanks again :)

**Comment by wez** [comment]
Each of those libraries have been modified, some of them quite heavily, to meet our needs and get ported to our C++ code base, so it is not possible to link against a system library and still operate correctly.

That aside, what value would factoring out these libraries provide, and to whom?  The watchman binary is a self contained program, so there is no risk of symbol collision that might be encountered if it was linked into some other program.

**Comment by hogarthj** [comment]
See debian policy section 4.13 [here](https://www.debian.org/doc/debian-policy/ch-source.html) and the Fedora [bundling page](https://fedoraproject.org/wiki/Bundled_Libraries) with the [relevant section of Fedora policy](https://fedoraproject.org/wiki/Packaging:Guidelines#Statically_Linking_Executables).

There are good reasons to unbundle libraries and to use system libraries, and all distributions have a preference to do so where at all possible.

The OP would like to see this application within Fedora, and it would be a useful addition to have in our repositories, but this is a [problem for the review](https://bugzilla.redhat.com/show_bug.cgi?id=1450590).

**Comment by wez** [comment]
Ah, that context is helpful.  I've previously commented on acceptance into distributions [here](https://github.com/facebook/watchman/issues/30#issuecomment-39049000).  The TL;DR is that we're happy to triage and figure out how best to make that happen, but we won't necessarily bend to every request or do the work to make that happen.

With that in mind; let's figure out how to solve this problem!

Could you tell me more about what specifically is the blocker here?  The modules mentioned in the OP are now effectively internal modules that we compile as static libraries and link into our final executable.  It's not clear from skimming the bugzilla issue you reference whether the linkage is a separate issue from the bundling.

Here's some context on each of the modules mentioned in the OP:

- jansson: modified to support BSER templating, then ported to C++.  There is no effective upstream at this point, because of both of these differences.  We're going to migrate to folly at some point in the future (this is gated by reducing the boost dependencies in folly).
- wildmatch didn't exist as a standalone library, and the code has no formal upstream.
- tap: modified for our C++ environment and likely to be replaced with something that is a better C++ fit in the future
- libart: significant changes for C++.  The upstream targets C99.  I have a patch series that will replace this once I've resolved some perf issues.

What are acceptable resolutions for this, given the above?

**Comment by wez** [comment]
@asymptotically @hogarthj any feedback?

**Comment by hogarthj** [comment]
@wez apologies for the delay, had some $life issues come up...

I've requested clarification on this part of the Fedora guidelines that define our policy here:

https://pagure.io/packaging-committee/issue/692

When I get a response from the FPC I can then follow this up.

In the spirit of the unbundling discussions we had I think this should be acceptable... but it'll be best to get confirmation.

**Comment by hogarthj** [comment]
So the FPC agrees with the spirit with it not being a blocker.

Quick question for you... 

I see https://pypi.python.org/pypi/pywatchman exists with a link back to this project, but the last upload is quite a long time ago.

If someone wanted to build on this watchman package with python2-pywatchman and python3-pywatchman packages is pypi a valid source for that still, or will some careful scripting need to be applied to separate it from the source of the main watchman here?

**Comment by wez** [comment]
The source of truth for pywatchman is this repo.  I'm in the middle of rolling a new release internally which I'll follow up with a public version bump and I'll update the package in pypi.

Python 3 support requires the version of pywatchman in this repo.

I'd recommend that folks packaging watchman also package up pywatchman and ship that at the same time, taking the version number from the watchman release (that's what we do with our internal package rolls), and ignore the pypi package.

So something like `watchman-4.9.0` for the main package, then `python2-pywatchman-4.9.0` and `python3-pywatchman-4.9.0` to hold the broken out python specific portions.

**Comment by hogarthj** [comment]
Hmm okay, I'll have to revisit this later as we have special guidelines for the packaging of python modules and our macros won't work with this setup.

The main watchman application I don't see any blocker for now though.

I'll finish off his review on the next few days

**Comment by wez** [comment]
(re-labeling this issue.  I'm inclined to close this out, but am happy to keep this around as a discussion thread for this topic)

**Comment by wez** [comment]
I'm going to close this out; please open issues for what we need to address as you find them!
